File: MockLibAndroid/SmartMockLib/src/main/java/com/crescentflare/smartmock/utility/SmartMockParamMatcher.java
package com.crescentflare.smartmock.utility;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

/**
 * Smart mock library utility: match parameters with wildcard support
 */
public class SmartMockParamMatcher
{
    /**
     * Private constructor, only static methods allowed
     */

    private SmartMockParamMatcher()
    {
    }


    /**
     * String or JSON matching
     */

    public static boolean deepEquals(JSONObject requireObject, JSONObject haveObject)
    {
        List<String> wantKeys = new ArrayList<>();
        Iterator<String> requireObjIterator = requireObject.keys();
        while (requireObjIterator.hasNext())
        {
            wantKeys.add(requireObjIterator.next());
        }
        for (String key : wantKeys)
        {
            if (!haveObject.has(key))
            {
                return false;
            }
            else if (haveObject.opt(key) instanceof JSONObject && requireObject.opt(key) instanceof JSONObject)
            {
                if (!deepEquals(requireObject.optJSONObject(key), haveObject.optJSONObject(key)))
                {
                    return false;
                }
            }
            else if (!paramEquals(requireObject.optString(key, ""), haveObject.optString(key, "")))
            {
                return false;
            }
        }
        return true;
    }

    public static boolean paramEquals(String requireParam, String haveParam)
    {
        if (haveParam == null)
        {
            return false;
        }
        String[] patternSet = requireParam.split("\\*", -1);
        if (patternSet.length == 0)
        {
            return true;
        }
        if (patternSet[0].length() > 0 && !patternEquals(safeSubstring(haveParam, 0, patternSet[0].length()), patternSet[0]))
        {
            return false;
        }
        return searchPatternSet(haveParam, patternSet) >= 0;
    }


    /**
     * Internal pattern matching
     */

    private static int searchPatternSet(String value, String[] patternSet)
    {
        while (patternSet.length > 0 && patternSet[0].length() == 0)
        {
            patternSet = Arrays.copyOfRange(patternSet, 1, patternSet.length);
        }
        if (patternSet.length == 0)
        {
            return 0;
        }
        int startPos = 0, pos;
        boolean searching;
        do
        {
            searching = false;
            pos = searchPattern(safeSubstring(value, startPos), patternSet[0]);
            if (pos >= 0)
            {
                if (patternSet.length == 1)
                {
                    if (startPos + pos + patternSet[0].length() == value.length())
                    {
                        return startPos + pos;
                    }
                }
                else
                {
                    int nextPos = startPos + pos + patternSet[0].length();
                    int setPos = searchPatternSet(safeSubstring(value, nextPos), Arrays.copyOfRange(patternSet, 1, patternSet.length));
                    if (setPos >= 0)
                    {
                        return startPos + pos;
                    }
                }
                startPos += pos + 1;
                searching = true;
            }
        }
        while (searching);
        return -1;
    }

    private static int searchPattern(String value, String pattern)
    {
        if (pattern.length() == 0)
        {
            return 0;
        }
        for (int i = 0; i < value.length(); i++)
        {
            if (pattern.charAt(0) == '?' || value.charAt(i) == pattern.charAt(0))
            {
                if (patternEquals(safeSubstring(value, i, i + pattern.length()), pattern))
                {
                    return i;
                }
            }
        }
        return -1;
    }

    private static boolean patternEquals(String value, String pattern)
    {
        if (value.length() != pattern.length())
        {
            return false;
        }
        for (int i = 0; i < pattern.length(); i++)
        {
            if (pattern.charAt(i) != '?' && value.charAt(i) != pattern.charAt(i))
            {
                return false;
            }
        }
        return true;
    }

    /**
     * String helper
     */

    private static String safeSubstring(String string, int start)
    {
        if (start > string.length())
        {
            return "";
        }
        return string.substring(start);
    }

    private static String safeSubstring(String string, int start, int end)
    {
        if (start > string.length())
        {
            start = string.length();
        }
        if (end > string.length())
        {
            end = string.length();
        }
        if (end <= start)
        {
            return "";
        }
        return string.substring(start, end);
    }
}

File: MockLibIOS/SmartMockLib/Classes/utility/SmartMockParamMatcher.swift
//
//  SmartMockParamMatcher.swift
//  SmartMockServer Pod
//
//  Main library utility: match parameters with wildcard support
//

class SmartMockParamMatcher {
    
    // --
    // MARK: Initialization
    // --
    
    private init() {
        // Private constructor, only static methods allowed
    }


    // --
    // MARK: String or JSON matching
    // --
    
    static func deepEquals(requireDictionary: [String: AnyObject], haveDictionary: [String: AnyObject]) -> Bool {
        let wantKeys = Array(requireDictionary.keys)
        for key in wantKeys {
            if !haveDictionary.keys.contains(key) {
                return false
            } else if haveDictionary[key] is [String: AnyObject] && requireDictionary[key] is [String: AnyObject] {
                if !deepEquals(requireDictionary: requireDictionary[key] as! [String: AnyObject], haveDictionary: haveDictionary[key] as! [String: AnyObject]) {
                    return false
                }
            } else {
                var requireParam = ""
                var haveParam = ""
                if requireDictionary[key] is String {
                    requireParam = requireDictionary[key] as! String
                } else if requireDictionary[key] is Int {
                    requireParam = String(requireDictionary[key] as! Int)
                } else if requireDictionary[key] is Bool {
                    requireParam = String(requireDictionary[key] as! Bool)
                }
                if haveDictionary[key] is String {
                    haveParam = haveDictionary[key] as! String
                } else if haveDictionary[key] is Int {
                    haveParam = String(haveDictionary[key] as! Int)
                } else if haveDictionary[key] is Bool {
                    haveParam = String(haveDictionary[key] as! Bool)
                }
                if !paramEquals(requireParam: requireParam, haveParam: haveParam) {
                    return false
                }
            }
        }
        return true
    }

    static func paramEquals(requireParam: String, haveParam: String?) -> Bool {
        if haveParam == nil {
            return false
        }
        let patternSet = requireParam.split(omittingEmptySubsequences: false, whereSeparator: { $0 == "*" }).map(String.init)
        if patternSet.count == 0 {
            return true
        }
        if patternSet[0].count > 0 && !patternEquals(value: safeSubstring(haveParam!, start: 0, end: patternSet[0].count), pattern: patternSet[0]) {
            return false
        }
        return searchPatternSet(value: haveParam!, paramPatternSet: patternSet) >= 0
    }

    
    // --
    // MARK: Internal pattern matching
    // --
    
    private static func searchPatternSet(value: String, paramPatternSet: [String]) -> Int {
        var patternSet = paramPatternSet
        while patternSet.count > 0 && patternSet[0].count == 0 {
            patternSet = Array(patternSet[1..<patternSet.count])
        }
        if patternSet.count == 0 {
            return 0
        }
        var startPos = 0
        var pos = 0
        var searching = false
        repeat {
            searching = false
            pos = searchPattern(value: safeSubstring(value, start: startPos), pattern: patternSet[0])
            if pos >= 0 {
                if patternSet.count == 1 {
                    if startPos + pos + patternSet[0].count == value.count {
                        return startPos + pos
                    }
                } else {
                    let nextPos = startPos + pos + patternSet[0].count
                    let setPos = searchPatternSet(value: safeSubstring(value, start: nextPos), paramPatternSet: Array(patternSet[1..<patternSet.count]))
                    if setPos >= 0 {
                        return startPos + pos
                    }
                }
                startPos += pos + 1
                searching = true
            }
        } while (searching)
        return -1
    }

    private static func searchPattern(value: String, pattern: String) -> Int {
        if pattern.count == 0 {
            return 0
        }
        var valueIndex = value.startIndex
        for i in 0..<value.count {
            if pattern[pattern.startIndex] == "?" || value[valueIndex] == pattern[pattern.startIndex] {
                if patternEquals(value: safeSubstring(value, start: i, end: i + pattern.count), pattern: pattern) {
                    return i
                }
            }
            valueIndex = value.index(valueIndex, offsetBy: 1)
        }
        return -1
    }

    private static func patternEquals(value: String, pattern: String) -> Bool {
        if value.count != pattern.count {
            return false
        }
        var patternIndex = pattern.startIndex
        var valueIndex = value.startIndex
        for _ in 0..<pattern.count {
            if pattern[patternIndex] != "?" && value[valueIndex] != pattern[patternIndex] {
                return false
            }
            patternIndex = pattern.index(patternIndex, offsetBy: 1)
            valueIndex = value.index(valueIndex, offsetBy: 1)
        }
        return true
    }
    

    // --
    // MARK: String helper
    // --
    
    private static func safeSubstring(_ string: String, start: Int) -> String {
        if start > string.count {
            return ""
        }
        return String(string[string.index(string.startIndex, offsetBy: start)...])
    }
    
    private static func safeSubstring(_ string: String, start: Int, end: Int) -> String {
        var startPos = start
        var endPos = end
        if startPos > string.count {
            startPos = string.count
        }
        if endPos > string.count {
            endPos = string.count
        }
        if endPos <= startPos {
            return ""
        }
        return String(string[string.index(string.startIndex, offsetBy: startPos)..<string.index(string.startIndex, offsetBy: endPos)])
    }
    
}
